File: playwright-tests/tests/login-valid.spec.ts
import { test, expect } from "@playwright/test";
import { LoginPage } from "./pages/login-page";
import { WelcomePage } from "./pages/welcome-page";

test.describe("Valid Login Tests", () => {
  let loginPage: LoginPage;
  let welcomePage: WelcomePage;

  test.beforeEach(async ({ page }) => {
    loginPage = new LoginPage(page);
    welcomePage = new WelcomePage(page);
    await loginPage.visit();
  });

  test.describe("when login page loads", () => {
    test("displays the login form with all elements", async () => {
      await loginPage.isDisplayed();
      expect(await loginPage.getTitleText()).toBe("Sign In");
      await expect(loginPage.usernameField).toBeVisible();
      await expect(loginPage.passwordField).toBeVisible();
      await expect(loginPage.signinButton).toBeVisible();
      await expect(loginPage.signinButton).toHaveText("Sign In");
    });

    test("has the correct page title", async ({ page }) => {
      await expect(page).toHaveTitle(/Login/);
    });

    test("shows placeholder text in input fields", async () => {
      await expect(loginPage.usernameField).toHaveAttribute("placeholder", "Enter your username");
      await expect(loginPage.passwordField).toHaveAttribute("placeholder", "Enter your password");
    });

    test("has the password field masked", async () => {
      await expect(loginPage.passwordField).toHaveAttribute("type", "password");
    });
  });

  test.describe("when logging in with valid credentials (admin)", () => {
    test("successfully logs in and shows the welcome page", async () => {
      await loginPage.login("admin", "admin123");

      await welcomePage.isDisplayed();
      expect(await welcomePage.getTitleText()).toBe("Welcome!");
      expect(await welcomePage.isLoggedInAs("admin")).toBe(true);
    });
  });

  test.describe("when logging in with valid credentials (testuser)", () => {
    test("successfully logs in and shows the welcome page", async () => {
      await loginPage.login("testuser", "Test@1234");

      await welcomePage.isDisplayed();
      expect(await welcomePage.getTitleText()).toBe("Welcome!");
      expect(await welcomePage.isLoggedInAs("testuser")).toBe(true);
    });
  });

  test.describe("when logging out after successful login", () => {
    test("returns to the login page after logout", async () => {
      await loginPage.login("admin", "admin123");
      await welcomePage.isDisplayed();

      await welcomePage.clickLogout();

      await loginPage.isDisplayed();
      expect(await loginPage.getTitleText()).toBe("Sign In");
    });

    test("clears the form fields after logout", async () => {
      await loginPage.login("admin", "admin123");
      await welcomePage.isDisplayed();

      await welcomePage.clickLogout();

      await expect(loginPage.usernameField).toHaveValue("");
      await expect(loginPage.passwordField).toHaveValue("");
    });
  });

  test.describe("when entering text into form fields", () => {
    test("accepts text in the username field", async () => {
      await loginPage.enterUsername("testuser");
      await expect(loginPage.usernameField).toHaveValue("testuser");
    });

    test("accepts text in the password field", async () => {
      await loginPage.enterPassword("mypassword");
      await expect(loginPage.passwordField).toHaveValue("mypassword");
    });
  });
});

File: ruby-watir-tests/spec/login_valid_spec.rb
require_relative "spec_helper"

RSpec.describe "Valid Login Tests" do
  let(:login_page) { LoginPage.new(@browser) }
  let(:welcome_page) { WelcomePage.new(@browser) }

  before(:each) do
    login_page.visit
  end

  context "when login page loads" do
    it "displays the login form with all elements" do
      expect(login_page).to be_displayed
      expect(login_page.title_text).to eq("Sign In")
      expect(login_page.username_field).to be_present
      expect(login_page.password_field).to be_present
      expect(login_page.signin_button).to be_present
      expect(login_page.signin_button.text).to eq("Sign In")
    end

    it "has the correct page title" do
      expect(@browser.title).to include("Login")
    end

    it "shows placeholder text in input fields" do
      expect(login_page.username_field.attribute_value("placeholder")).to eq("Enter your username")
      expect(login_page.password_field.attribute_value("placeholder")).to eq("Enter your password")
    end

    it "has the password field masked" do
      expect(login_page.password_field.attribute_value("type")).to eq("password")
    end
  end

  context "when logging in with valid credentials (admin)" do
    it "successfully logs in and shows the welcome page" do
      login_page.login("admin", "admin123")

      expect(welcome_page).to be_displayed
      expect(welcome_page.title_text).to eq("Welcome!")
      expect(welcome_page.logged_in_as?("admin")).to be true
    end
  end

  context "when logging in with valid credentials (testuser)" do
    it "successfully logs in and shows the welcome page" do
      login_page.login("testuser", "Test@1234")

      expect(welcome_page).to be_displayed
      expect(welcome_page.title_text).to eq("Welcome!")
      expect(welcome_page.logged_in_as?("testuser")).to be true
    end
  end

  context "when logging out after successful login" do
    it "returns to the login page after logout" do
      login_page.login("admin", "admin123")

      expect(welcome_page).to be_displayed

      welcome_page.click_logout

      # Should be back on login page
      expect(login_page).to be_displayed
      expect(login_page.title_text).to eq("Sign In")
    end

    it "clears the form fields after logout" do
      login_page.login("admin", "admin123")
      expect(welcome_page).to be_displayed

      welcome_page.click_logout

      expect(login_page.username_field.value).to eq("")
      expect(login_page.password_field.value).to eq("")
    end
  end

  context "when entering text into form fields" do
    it "accepts text in the username field" do
      login_page.enter_username("testuser")
      expect(login_page.username_field.value).to eq("testuser")
    end

    it "accepts text in the password field" do
      login_page.enter_password("mypassword")
      expect(login_page.password_field.value).to eq("mypassword")
    end
  end
end
